crates/adapters/src/transport/s3/mod.rs

```diff
@@ -446,21 +451,88 @@ impl S3InputReader {
                                 }
                             }
                             if added_chunks {
-                                queue.back_mut().unwrap().end_offset = None;
+                                queue.lock().await.back_mut().unwrap().end_offset = None;
                             }
                         }
-                        None => {
-                            // End of input.
-                            consumer.eoi();
-                            running = false;
+                        Err(_) => break, // Channel closed
+                    }
+                }
+            });
+            join_handles.push(handle);
+        }
+
+        // Wait for end of input.
+        let consumer_clone = consumer.clone();
+        tokio::task::spawn(async move {
+            _ = join_all(join_handles).await;
```

Comment:
I had no idea this was valid syntax, I thought you had to write `let _ = ...`. But the compiler accepts it without warning.

Reply:
I didn't even notice. I also didn't realize this was a thing.